cranlock/lock.py

```diff
@@ -0,0 +1,226 @@
+
+
+def add_to_graph(graph: dict, dependency_tree):
+    """Add a dependency tree to our depedency graph recursively. This truns the tree into a
+    (hopefully) directed acyclic graph that we can sort using a depth-first search.
+    """
+    if len(dependency_tree.keys()) == 0:
+        return
+
+    for dep, nesteds in dependency_tree.items():
+
+        if graph.get(dep, None) is None:
+            graph[dep] = set()
+
+        if nesteds == []:
+            continue
+
+        for nested in nesteds.keys():
+            graph[dep].add(nested)
+
+        add_to_graph(graph, nesteds)
+
+
+def get_dependency_graph(packages: [str]):
+    """Get the dependency graph for a list of packages by getting the tree for each one then
+    adding them to a graph using `add_to_graph`"""
+    dependencies = dict(map(lambda package: (package, get_all_dependencies(package)), packages))
+
+    graph = {}
+    add_to_graph(graph, dependencies)
+    return graph
+
+
+class Mark(Enum):
+    Permanent = 0
+    Temporary = 1
+    Empty = 2
+
+
+def have_visited_all(visited: dict) -> bool:
+    """Have we visited all the nodes in the graph?"""
+    for mark in visited.values():
+        if mark == Mark.Empty:
+            return False
+
+    return True
+
+
+def get_first_unvisited(visited: dict) -> str:
+    """Get the first unvisited node in the graph, or None if we've visited all of them"""
+    for (key, mark) in visited.items():
+        if mark == Mark.Empty:
+            return key
+
+    return None
+
+
+def visit(graph, node, visited, output):
+    """Visit a node in a depth-first search"""
+    mark = visited[node]
+    if mark == Mark.Permanent:
+        return
+    elif mark == Mark.Temporary:
+        # This should (hopefully) never happen because package dependencies *shouldn't* have cycles
+        raise Exception("Somehow, the dependency graph is not directed and acyclic")
+
+    visited[node] = Mark.Temporary
+
+    for child_node in graph[node]:
+        visit(graph, child_node, visited, output)
+
+    visited[node] = Mark.Permanent
+
+    # Append the nodes, rather than adding them to the head, because we want to install
+    # dependencies before their dependents
+    output.append(node)
+
+
+def sort_dependency_graph(graph: dict):
+    """Sort the dependency graph into a list using depth-first search"""
+    visited = dict.fromkeys(graph.keys(), Mark.Empty)
+
+    output = []
+
+    while not have_visited_all(visited):
+        node = get_first_unvisited(visited)
+
+        visit(graph, node, visited, output)
+
+        visited[node] = True
+
+    return output
+
+
+def extract_name_and_version(input_line: str) -> (str, str):
+    """Extract the name and version of a package from an input line from a versions.tsv file.
+    The file format is TSV, with the name in the first column and the version in the second
+    """
+    info = list(filter(lambda s: len(s) > 0, input_line.strip().split('\t')))
```

Comment:
Do you think you could wrap this lambda in a  set of `()`? It wasn't immediately obvious where it ended when I first looked at it.